File: sdk/python/examples/63c_run_by_name.py
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License. See LICENSE file in the project root for details.

"""Run by Name — execute a pre-deployed agent without its definition.

Demonstrates:
    - runtime.run("workflow_name", prompt) — run by name
    - runtime.start("workflow_name", prompt) — fire-and-forget by name
    - runtime.stream("workflow_name", prompt) — stream by name
    - Optional version parameter for workflow versioning

The agent must be deployed (63_deploy.py) and workers must be
running (63b_serve.py) for tool tasks to execute.

Requirements:
    - Conductor server running
    - Agent deployed (run 63_deploy.py first)
    - Workers running (run 63b_serve.py in another terminal)
    - AGENTSPAN_SERVER_URL=http://localhost:8080/api in .env or environment
"""

from agentspan.agents import AgentRuntime

with AgentRuntime() as runtime:
    # ── Run by name (synchronous, blocks until complete) ─────────────
    print("Running doc_assistant by name...")
    result = runtime.run("doc_assistant", "How do I reset my password?")
    print(f"Output: {result.output}\n")

    # ── Start by name (fire-and-forget, returns handle) ──────────────
    print("Starting ops_bot by name...")
    handle = runtime.start("ops_bot", "Check the status of the API gateway")
    print(f"Started workflow: {handle.workflow_id}")

    # Wait for it to complete
    status = handle.get_status()
    print(f"Status: {status.status}")
    if status.is_complete:
        print(f"Output: {status.output}\n")

    # ── Stream by name ───────────────────────────────────────────────
    print("Streaming doc_assistant by name...")
    for event in runtime.stream("doc_assistant", "What are the API rate limits?"):
        if event.type == "token":
            print(event.token, end="", flush=True)
        elif event.type == "done":
            print(f"\n\nFinal output: {event.output}")

File: sdk/typescript/examples/63c-run-by-name.ts
/**
 * 63c - Run by Name — execute a pre-deployed agent without its definition.
 *
 * Demonstrates the concept of running a deployed agent by workflow name.
 *
 * NOTE: The TypeScript SDK's runtime.run() currently requires an Agent
 * object. To run by name, use the HTTP API directly:
 *   POST /api/agent/run { "name": "agent_doc_assistant", "prompt": "..." }
 *
 * Requirements:
 *   - Conductor server running
 *   - Agent deployed (run 63-deploy.ts first)
 *   - Workers running (run 63b-serve.ts in another terminal)
 *   - AGENTSPAN_SERVER_URL=http://localhost:8080/api as environment variable
 */

// Only run when executed directly (not when imported for discovery)
if (process.argv[1]?.endsWith('63c-run-by-name.ts') || process.argv[1]?.endsWith('63c-run-by-name.js')) {
  console.log('Run by Name -- Workflow Name Execution');
  console.log('');
  console.log('The TypeScript SDK currently requires an Agent object for runtime.run().');
  console.log('To run a deployed agent by name, use the HTTP API directly:');
  console.log('');
  console.log('  // Run by name (synchronous)');
  console.log('  POST /api/agent/run');
  console.log('  { "name": "agent_doc_assistant", "prompt": "How do I reset my password?" }');
  console.log('');
  console.log('  // Start by name (fire-and-forget)');
  console.log('  POST /api/agent/start');
  console.log('  { "name": "agent_ops_bot", "prompt": "Check the status of the API gateway" }');
  console.log('');
  console.log('  // Stream by name');
  console.log('  GET /api/agent/stream/{executionId}');
  console.log('');
  console.log('Or use the Python SDK which supports runtime.run("workflow_name", prompt).');
}
